crates/noirc_frontend/src/hir/def_collector/dc_mod.rs

```diff
@@ -116,6 +253,104 @@ impl<'a> ModCollector<'a> {
+                    errors.push(error.into_file_diagnostic(self.file_id));
+                }
+            }
+        }
+    }
+
+    fn get_unresolved_trait(&self, module_def_id: ModuleDefId) -> Option<&UnresolvedTrait> {
+        match module_def_id {
+            ModuleDefId::TraitId(trait_id) => self.def_collector.collected_traits.get(&trait_id),
+            _ => None,
+        }
+    }
+
+    fn collect_trait_from_module(
+        &mut self,
+        context: &mut Context,
+        module_def_id: ModuleDefId,
+        trait_impl: TraitImpl,
+        errors: &mut Vec<FileDiagnostic>,
+        trait_name: &Ident,
+    ) {
+        if let Some(collected_trait) = self.get_unresolved_trait(module_def_id) {
+            let trait_def = collected_trait.trait_def.clone();
+            let impl_type_span = trait_impl.object_type_span;
+            let impl_generics = trait_impl.impl_generics.clone();
+            let impl_object_type = trait_impl.object_type.clone();
+            let collected_implementations =
+                self.collect_trait_implementations(context, &trait_impl, &trait_def, errors);
+            let key = (impl_object_type, self.module_id);
+            self.def_collector.collected_traits_impls.entry(key).or_default().push((
+                impl_generics,
+                impl_type_span,
+                collected_implementations,
+            ));
+        } else {
+            let error = DefCollectorErrorKind::NotATrait {
+                primary_message: format!(
+                    "{} is not a trait, therefore it can't be implemented",
+                    trait_name
+                ),
+                secondary_message: "".to_string(),
+                span: trait_name.span(),
+            };
+            errors.push(error.into_file_diagnostic(self.file_id));
+        }
+    }
+
+    fn collect_trait_implementations(
+        &mut self,
+        context: &mut Context,
+        trait_impl: &TraitImpl,
+        trait_def: &NoirTrait,
+        errors: &mut Vec<FileDiagnostic>,
+    ) -> UnresolvedFunctions {
+        let mut unresolved_functions =
+            UnresolvedFunctions { file_id: self.file_id, functions: Vec::new() };
+        for item in &trait_impl.items {
+            if let TraitImplItem::Function(noir_function) = item {
+                match check_trait_method_implementation(trait_def, &noir_function) {
+                    Ok(()) => {
+                        let func_id = context.def_interner.push_empty_fn();
+                        context
+                            .def_interner
+                            .push_function_definition(noir_function.name().to_owned(), func_id);
+                        unresolved_functions.push_fn(self.module_id, func_id, noir_function.clone());
+                    }
+                    Err(error) => {
+                        errors.push(error.into_file_diagnostic(self.file_id));
+                    }
+                }
+            } else {
+            }
```

Comment:
Should we add an unimplemented warning for the other items?

Reply:
No, the other items will be check in other places. I will delete the `else` part